src/xb-value-bindings.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (C) 2020 Endless OS Foundation LLC
+ *
+ * Author: Philip Withnall <withnall@endlessm.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN				"XbValueBindings"
+
+#include "config.h"
+
+#include <glib.h>
+
+#include "xb-opcode-private.h"
+#include "xb-value-bindings.h"
+
+typedef struct {
+	enum {
+		KIND_NONE,
+		KIND_TEXT,
+		KIND_INTEGER,
+	} kind;
+	union {
+		gchar *text;
+		guint32 integer;
+	};
+	GDestroyNotify destroy_func;
+} BoundValue;
+
+typedef struct {
+	/* Currently limited to two values since that’s all that any client
+	 * uses. This could be expanded to dynamically allow more in future. */
+	BoundValue values[2];
```

Comment:
That seems a bit odd to hardcode here

Reply:
It’s cutting a corner, but saved me having to implement dynamic sizing for the `XbValueBindings` struct, which isn’t necessary at the moment. There are actually very few queries in gnome-software or fwupd which use bound values, and they only ever use 2 at the most.

Reply:
I guess that's pragmatic. Is it in the docs anywhere? Maybe we could make it say 8 instead as I can think offhand of a place in fwupd that I could use this with three bound values.

Reply:
I was thinking that rather than documenting it, we’d just fix it to expand dynamically if someone hits the limit. Which approach would you prefer?

Reply:
I’ve expanded it to 4 values (from 2), but don’t want to go as far as 8 because that will make the struct absolutely massive. There’s now plenty of room in there for rearranging things in future to use dynamic values, without breaking ABI.